app/controllers/users/nbp_wallet_controller.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+module Users
+  class NbpWalletController < ApplicationController
+    CONNECTOR_URL = 'http://localhost:80'
+    API_KEY = 'myapikey'
+    DISPLAY_NAME = 'CodeHarbor'
+    REQUIRED_ATTRIBUTES = %w[GivenName Surname EMailAddress].freeze # https://enmeshed.eu/integrate/attribute-values
+
+    skip_before_action :require_user!
+    skip_after_action :verify_authorized
+    before_action :set_provider_details
+    before_action :set_connector_api
+
+    def connect; end
+
+    def generate_qr_code
+      if session[:relationship_template_id].present?
+        reject_all_relationships(session[:relationship_template_id])
+      end
+      template_id = @connector.create_relationship_template(DISPLAY_NAME, REQUIRED_ATTRIBUTES)
+      session[:relationship_template_id] = template_id
+
+      render body: @connector.template_qr_code(template_id), content_type: 'image/png'
+    end
+
+    def relationship_status
+      render json: {status: 'missing relationship_template_id'} and return if session[:relationship_template_id].nil?
+
+      relationships = @connector.pending_relationships_for_rlt(session[:relationship_template_id])
+
+      render json: {status: 'waiting'} and return if relationships.empty?
+
+      render json: {status: 'ready'}
+    end
+
+    def finalize
+      relationships = @connector.pending_relationships_for_rlt(session[:relationship_template_id])
+      abort_and_refresh and return if relationships.size != 1
+
+      accept_and_create_user(relationships.first)
+    end
+
+    private
+
+    def accept_and_create_user(relationship)
+      relationship_changes = relationship['changes']
+
+      abort_and_refresh if relationship_changes.size != 1
+
+      return unless create_user(@connector.parse_userdata_in_rel_change(relationship_changes.first))
+
+      if @connector.accept_relationship_change(relationship['id'], relationship_changes.first['id'])
+        sign_in_and_redirect @user, event: :authentication
+      else
+        abort_and_refresh
+        @user.destroy!
+      end
+    end
+
+    def create_user(userdata)
+      @user = User.new(
+        first_name: userdata['GivenName'],
+        last_name: userdata['Surname'],
+        email: userdata['EMailAddress'],
+        password: Devise.friendly_token[0, 20],
+        identities: [UserIdentity.new(provider_uid: @provider_uid, omniauth_provider: @provider)]
+      )
+      @user.skip_confirmation! # TODO: check email
+      @user.save
```

Comment:
Yes, I see that. The one advantage I assume is to ensure some compatibility with further refactorings. My goal is to make the differences with the Wallet sign in as small as possible, so that we can re-use many of the existing structures.

Reply:
Maybe we can even make the "mapping" even more "explicit"?

Reply:
After the refactoring of ```omniauth_callbacks_controller.rb``` and ```User.from_omniauth``` I decided to inline the method.